src/components/views/rooms/wysiwyg_composer/utils/mentions.ts

```diff
@@ -0,0 +1,53 @@
+/*
+Copyright 2023 The Matrix.org Foundation C.I.C.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+import { parsePermalink } from "../../../../../utils/permalinks/Permalinks";
+
+// a rich text link looks like <a href="href"...>link text</a>
+const richTextLinkRegex = /(<a.*?<\/a>)/g;
+
+// a markdown link looks like [link text](<href>), regex uses lookahead and negation to avoid super-linear performance
+const mdLinkRegex = /\[(?=([^\]]*))\1\]\(<(?=([^>]*))\2>\)/g;
```

Comment:
I realise you've spent a while on this, but...

Can you explain what purpose the lookahead assertions are serving here? As I understand it, focussing on the `(?=([^\]]*))\1` bit:

* `(?=([^\]]*))` matches if the next characters are a sequence of characters other than `]`, but does not actually consume those characters
* `\1` then matches the same thing as the above, and actually consumes it

... so how is that different to just `([^\]]*)` ?

Reply:
I got to this using the [Sonar cloud advice page. ](https://rules.sonarsource.com/typescript/RSPEC-5852). Specifically I looked at the "Compliant solution" section.

It unhelpfully tells me that what I want is a something that isn't available in JS, namely possessive quantifiers. From reading around though, it seems that by using the lookahead I can achieve the aim of having a something that won't give back anything that they've matched. As they won't give back anything they've matched, this reduces the backtracking. 

Apologies if the terminology is off - this is quite new to me.

Reply:
oh I see. That does make some sense.

I think it would be good to expand the comment to explain some of this. I'd say something like...

> We need to stop the regex backtracking, to avoid super-linear performance. To do this, we use a lookahead assertion `(?=...)`, which greedily matches up to the closing `]`, and then match the contents of the assertion with a backreference `\1`. Since the backreference matches something that has already matched, it will not backtrack.